docs/development.md

```diff
@@ -0,0 +1,775 @@
+        /relationships
+    /phase-02-build-relationships
+        /relationships
+          8fcc6865-817d-4952-ac53-8248b357b5d8.json
+```
+
+Integration data that is staged for publishing will be stored under the
+`.j1-integration/graph` directory. Files containing Entities will be suffixed
+with `.entities.json` and files containing Relationships will be suffixed with
+`.relationships.json`. This is done because the directory structure of the
+`graph` directory is meant to assist with debugging and provide developrs
+insight about the data collect in each phase. During synchronization, the
+directory will be blind
+
+Data will be indexed by `_type` and stored under the
+`.j1-integration/index/_type` directory.
+
+Each `json` file will store data under the following format (describe with a
+typescript interface):
+
+```
+interface FlushedEntityGraphData {
+  entities: Entity[]
+}
+
+interface FlushedRelationshipGraphData {
+  relationships: Relationship[]
+}
+```
+
+It's best to keep a "collect and forget" mindset to avoid retaining too much
+collected data in memory. Flushed data can always be pulled back into memory
+later via the `listEntitiesByType` and `listRelationshipsByType` and reading
+data from disk is decently fast.
+
+#### Failure handling
+
+By default, the framework will only halt the execution of an integration if the
+configuration validation fails. Failures that occur during the execution of a
+phase or step will not halt the execution of later phases. Information on which
+steps and phases that have failed will be collected and published as metadata
+when performing synchronization with JupiterOne. A failure in a step or phase
+will automatically be logged along with context about the error that occurred.
+At the end of an integration run, a summary will be displayed of the overall
+status to give developers a good idea of how failures will have affected the
+JupiterOne graph.
+
+An example summary of the phases and steps will like this:
+
+```json
+[
+  {
+    "id": "account-phase",
+    "name": "Collect accounts",
+    "status": "success"
+  },
+  {
+    "id": "users-and-groups-phase",
+    "name": "Collect users and groups",
+    "status": "partial_failure",
+    "steps": [
+      {
+        "id": "fetch-users",
+        "name": "Fetch Users",
+        "types": ["my_integration_user"],
+        "status": "failure"
+      },
+      {
+        "id": "fetch-groups",
+        "name": "Fetch Groups",
+        "types": ["my_integration_group"],
+        "status": "success"
+      }
+    ]
+  },
+  {
+    "id": "build-relationships-phase",
+    "name": "Collect accounts",
+    "status": "partial_success",
+    "steps": [
+      {
+        "id": "build-user-to-group-relationships",
+        "name": "Fetch Accounts",
+        "types": ["my_integration_user_to_group_relationship"],
+        "dependsOn": ["fetch-users", "fetch-groups"],
+        "status": "partial_success_from_dependency_failure"
```

Comment:
It would be nice to have a list of possible phase and step exit statuses in this section.

Reply:
will add